### DISCUSSION TITLE
Templetazing typedefs / slices

### DISCUSSION THREAD

**Discussion by Nizarll** [c3lang/c3c#1562] [Ideas]
documentation doesnt seem to show anything like this so maybe
allowing generic types on slices and defs kind of like this would be a great idea or just move away generics out of modules to another form of syntax so that structs and defs can be generalised without having to create a new module: 
```c3
module raylib::rect<T>;
def Rect = T[<4>];
```

**Comment by lerno** [comment]
If this is what you mean, then it already works: https://godbolt.org/z/1Y5Wjj9fK

**Reply by Nizarll** [reply]
you cannot do that for simple arrays only slices which causes a problem for web assembly since interfacing with javascript functions unwraps  slices as different arguments instead of converting it to a normal array

**Reply by lerno** [reply]
What do you mean "you can't do that"? The godbolt example is showing vectors, but it also works on arrays of course, or slices. Can you show me the code that doesn't work?

**Reply by Nizarll** [reply]
ok i figured it out after trying some things out a bit more since there is no way to delimit modules you have to leave all submodules ( even the generic structures and types ) and the end of the file so it doesnt conflict with all the code of the module itself which is why it wasnt compiling

**Reply by lerno** [reply]
What do you mean there is no way to delimit modules?

**Reply by Nizarll** [reply]
like each module ends where a new module is found so if i do something like this : https://godbolt.org/z/PT95j6Trc doesnt work

**Reply by lerno** [reply]
Yes, placing the main method in a generic module is going to be a bad thing. But why not just invert the order: https://godbolt.org/z/Enc8d11vT
